### PR TITLE
HPR-1172: Add `hcp_packer_channel_assignment` Resource

### PR DESCRIPTION
**Pull request by aidan-mundy**
### :hammer_and_wrench: Description

- Adds `hcp_packer_channel_assignment` resource
- `hcp_packer_channel` resource 
  - Deprecates `iteration` block
  - Adds read-only `restricted` field
- Minor dev dependency changes
  - Updates mongo-driver to latest patch version because current version was retracted
  - Updates config for `depguard` to no longer break on newer versions of `golangci-lint`

### :building_construction: Acceptance tests

- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAcc.*Packer.* -test.v'
==> Checking that code complies with gofmt requirements...
golangci-lint run --config ./golangci-config.yml 
TF_ACC=1 go test ./internal/... -v -run=TestAcc.*Packer.* -test.v -timeout 360m -parallel=10
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/clients    (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/consul     (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/input      (cached) [no tests to run]
=== RUN   TestAcc_dataSourcePacker
--- PASS: TestAcc_dataSourcePacker (7.11s)
=== RUN   TestAcc_dataSourcePacker_revokedIteration
--- PASS: TestAcc_dataSourcePacker_revokedIteration (10.96s)
=== RUN   TestAcc_dataSourcePackerImage
--- PASS: TestAcc_dataSourcePackerImage (7.20s)
=== RUN   TestAcc_dataSourcePackerImage_revokedIteration
--- PASS: TestAcc_dataSourcePackerImage_revokedIteration (10.34s)
=== RUN   TestAcc_dataSourcePackerImage_channelAndIterationIDReject
--- PASS: TestAcc_dataSourcePackerImage_channelAndIterationIDReject (3.57s)
=== RUN   TestAcc_dataSourcePackerImage_channelAccept
--- PASS: TestAcc_dataSourcePackerImage_channelAccept (7.28s)
=== RUN   TestAcc_dataSourcePackerIteration
--- PASS: TestAcc_dataSourcePackerIteration (5.68s)
=== RUN   TestAcc_dataSourcePackerIteration_revokedIteration
--- PASS: TestAcc_dataSourcePackerIteration_revokedIteration (11.19s)
=== RUN   TestAccPackerChannelAssignment_SimpleSetUnset
--- PASS: TestAccPackerChannelAssignment_SimpleSetUnset (14.99s)
=== RUN   TestAccPackerChannelAssignment_AssignLatest
--- PASS: TestAccPackerChannelAssignment_AssignLatest (16.30s)
=== RUN   TestAccPackerChannelAssignment_InvalidInputs
--- PASS: TestAccPackerChannelAssignment_InvalidInputs (6.11s)
=== RUN   TestAccPackerChannelAssignment_CreateFailsWhenPreassigned
--- PASS: TestAccPackerChannelAssignment_CreateFailsWhenPreassigned (8.84s)
=== RUN   TestAccPackerChannelAssignment_HCPManagedChannelErrors
--- PASS: TestAccPackerChannelAssignment_HCPManagedChannelErrors (3.48s)
=== RUN   TestAccPackerChannelAssignment_EnforceNull
--- PASS: TestAccPackerChannelAssignment_EnforceNull (30.68s)
=== RUN   TestAccPackerChannel
--- PASS: TestAccPackerChannel (6.49s)
=== RUN   TestAccPackerChannel_AssignedIteration
--- PASS: TestAccPackerChannel_AssignedIteration (8.29s)
=== RUN   TestAccPackerChannel_UpdateAssignedIteration
--- PASS: TestAccPackerChannel_UpdateAssignedIteration (13.32s)
=== RUN   TestAccPackerChannel_UpdateAssignedIterationWithFingerprint
--- PASS: TestAccPackerChannel_UpdateAssignedIterationWithFingerprint (6.47s)
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/provider   178.682s
```

### CODE REVIEW

File: docs/guides/packer-channel-management.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+---
+subcategory: ""
+page_title: "Advanced Packer Channel Management - HCP Provider"
+description: |-
+    A guide to integreting HCP Packer resources and data sources for more advanced channel management.
+---
+
+# Advanced Packer Channel Management
+
+By integrating multiple HCP Packer resources and data sources, you can perform more advanced channel management tasks.
```
Comment (nywilken) [inline docs/guides/packer-channel-management.md:10]: Slight reword to clarify the intent of the guide a bit. 
```suggestion
You can integrate multiple HCP Packer resources and data sources to perform advanced channel management tasks.
```

File: docs/resources/packer_channel.md [hunk 2]
```diff
@@ -1,80 +1,36 @@
...
-```
-
-
 <!-- schema generated by tfplugindocs -->
 ## Schema
 
 ### Required
 
-- `bucket_name` (String) The slug of the HCP Packer Registry image bucket where the channel should be created in.
+- `bucket_name` (String) The slug of the HCP Packer Registry image bucket where the channel should be created.
```
Comment (nywilken) [inline docs/resources/packer_channel.md:27]: Should we drop `image` here, as well?

```suggestion
- `bucket_name` (String) The slug of the HCP Packer Registry bucket where the channel should be created.
```

File: internal/provider/resource_packer_channel.go [hunk 3]
```diff
@@ -46,7 +45,7 @@ func resourcePackerChannel() *schema.Resource {
 				ValidateDiagFunc: validateSlugID,
 			},
 			"bucket_name": {
-				Description:      "The slug of the HCP Packer Registry image bucket where the channel should be created in.",
+				Description:      "The slug of the HCP Packer Registry image bucket where the channel should be created.",
```
Comment (nywilken) [inline internal/provider/resource_packer_channel.go:48]:
```suggestion
				Description:      "The slug of the HCP Packer Registry bucket where the channel should be created.",
```

File: internal/provider/resource_packer_channel_assignment.go [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+		return diag.FromErr(err)
+	}
+
+	bucketName := d.Get("bucket_name").(string)
+	channelName := d.Get("channel_name").(string)
+
+	iteration := &packermodels.HashicorpCloudPackerIteration{}
+	assignmentHasChanges := d.HasChanges("iteration_version", "iteration_id", "iteration_fingerprint")
+	if !assignmentHasChanges || d.HasChange("iteration_version") {
+		iteration.IncrementalVersion = int32(d.Get("iteration_version").(int))
```
Comment (nywilken) [inline internal/provider/resource_packer_channel_assignment.go:228]: I'm a little confused on what the expected value of iteration_version is meant to be here?

**Review by nywilken — COMMENTED** [review]
Took a first pass at the resource. Will continue looking through the import and run the acceptance locally.

File: internal/provider/resource_packer_channel_assignment.go (hunk 4 above)
Comment (aidan-mundy) [reply internal/provider/resource_packer_channel_assignment.go:228]: `iteration_version` contains the incremental version. Terraform stores this value as an `int`, but the iteration struct used throughout the code expects an `int32`

**Review by nywilken — APPROVED** [review]
This looks good to me. I'm running the tests locally. Approving in advance to not block. If all the test are good.

File: internal/provider/resource_packer_channel_assignment_test.go [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+			testAccCheckAssignmentStateMatchesIteration(assignment.ResourceName(), nil),
+			testAccCheckAssignmentStateMatchesChannelState(assignment.ResourceName(), channel.ResourceName()),
+			testAccCheckAssignmentStateMatchesAPI(assignment.ResourceName()),
+		)
+
+		return []resource.TestStep{
+			{ // Set up channel and set the assignment using Terraform
+				// This should be a no-op unless it is the first step, where the channel and null assignment are
+				// initially created. However, this step is included every time to make sure we've applied the
+				// assignment to the state at least once before checking if it is properly enforced against OOB changes.
```
Comment (nywilken) [inline internal/provider/resource_packer_channel_assignment_test.go:343]: Does OOB mean out of band?

Comment (aidan-mundy) [reply internal/provider/resource_packer_channel_assignment_test.go:343]: Yes

File: internal/provider/resource_packer_channel_assignment.go [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+		iteration = *i
+	}
+
+	if err := d.Set("iteration_version", iteration.IncrementalVersion); err != nil {
+		return err
+	}
+
+	id := iteration.ID
+	if id == "" {
+		id = "none"
```
[the remark this answers is not in the document]
Comment (aidan-mundy) [reply internal/provider/resource_packer_channel_assignment.go:443]: Will do

File: internal/provider/resource_packer_channel_assignment.go (hunk 4 above)
Comment (aidan-mundy) [reply internal/provider/resource_packer_channel_assignment.go:228]: Yup, you got it

File: internal/provider/resource_packer_channel_assignment.go [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+The ID of the HCP project where the channel is located. 
+If not specified, the project specified in the HCP Provider config block will be used, if configured.
+If a project is not configured in the HCP Provider config block, the oldest project in the organization will be used.`,
+				Type:         schema.TypeString,
+				Computed:     true,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+			"iteration_fingerprint": {
```
[the remark this answers is not in the document]
Comment (aidan-mundy) [reply internal/provider/resource_packer_channel_assignment.go:66]: It is not possible to run `terraform apply` unless one of the iteration identifier attributes has been set manually (and the two string attributes cannot be set to `""`). To unassign, the user can set one of the attributes to the "zero values" documented in the example for the resource.

File: internal/provider/resource_packer_channel_assignment.go [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+			Severity: diag.Error,
+			Summary:  fmt.Sprintf("HCP Packer channel with (channel_name %q) (bucket_name %q) (project_id %q) is managed by HCP Packer and cannot have an iteration assigned by Terraform.", channelName, bucketName, loc.ProjectID),
+		}}
+	} else if iteration := channel.Iteration; iteration != nil && (iteration.IncrementalVersion > 0 || iteration.ID != "" || iteration.Fingerprint != "") {
+		return diag.Diagnostics{diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  fmt.Sprintf("HCP Packer channel with (channel_name %q) (bucket_name %q) (project_id %q) already has an assigned iteration.", channelName, bucketName, loc.ProjectID),
+			Detail:   "To adopt this resource into Terraform, use `terraform import`, or remove the channel's assigned iteration using the HCP Packer GUI/API.",
+		}}
+	}
```
[the remark this answers is not in the document]
Comment (aidan-mundy) [reply internal/provider/resource_packer_channel_assignment.go:175]: I've chained these with else-if to create an implicit ordering requirement for the error checks. If the order of the Managed check and Iteration check is swapped, the user would receive an inappropriate error message that is still _technically_ correct, which is confusing. The err/channel checks and channel/managed checks already have implicit ordering by convention and practicality (check error before checking output, and check nil before checking a field), but I think it is clearer to string all of them together with if-else if you're already going to do it for two of them.

**Review by martinb3 — APPROVED** [review]

**Pull request merged**